File: ios/CombatTimer/TimerViewModel.swift
import Foundation

@MainActor
final class TimerViewModel: ObservableObject {
    @Published private(set) var uiState = TimerUiState()

    private let soundEngine = SoundEngine()
    private var tickTask: Task<Void, Never>?

    private static let tenSecondWarning = 10
    private static let restCountdownBeeps = 5
    private static let resumablePhases: Set<TimerPhase> = [.round, .rest, .goldenScore]

    /// Loads a sport, optionally overriding its configurable fields, ready to start.
    func configure(preset: SportPreset, rounds: Int? = nil, roundSeconds: Int? = nil, restSeconds: Int? = nil) {
        tickTask?.cancel()
        let resolvedRoundSeconds = roundSeconds ?? preset.roundSeconds
        uiState = TimerUiState(
            preset: preset,
            phase: .idle,
            totalRounds: rounds ?? preset.defaultRounds,
            currentRound: 1,
            roundSeconds: resolvedRoundSeconds,
            restSeconds: restSeconds ?? preset.restSeconds,
            secondsRemaining: resolvedRoundSeconds,
            isRunning: false
        )
    }

    func start() {
        if uiState.phase == .idle {
            uiState.phase = .round
            uiState.secondsRemaining = uiState.roundSeconds
        }
        runTicker()
    }

    func pause() {
        tickTask?.cancel()
        uiState.isRunning = false
    }

    func resume() {
        if Self.resumablePhases.contains(uiState.phase) {
            runTicker()
        }
    }

    func reset() {
        guard let preset = uiState.preset else { return }
        configure(
            preset: preset,
            rounds: uiState.totalRounds,
            roundSeconds: uiState.roundSeconds,
            restSeconds: uiState.restSeconds
        )
    }

    /// Called from the Judo "match tied" prompt after the final round buzzer.
    func startGoldenScore() {
        tickTask?.cancel()
        uiState.phase = .goldenScore
        uiState.secondsRemaining = 0
        uiState.isRunning = true
        runTicker()
    }

    /// Ends the match without golden score (decision reached in regulation, or golden score scored).
    func finishMatch() {
        tickTask?.cancel()
        uiState.phase = .finished
        uiState.isRunning = false
    }

    private func runTicker() {
        tickTask?.cancel()
        uiState.isRunning = true
        tickTask = Task { [weak self] in
            while true {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                if Task.isCancelled { return }
                await self?.tick()
            }
        }
    }

    private func tick() {
        switch uiState.phase {
        case .round: tickRound()
        case .rest: tickRest()
        case .goldenScore: uiState.secondsRemaining += 1
        default: break
        }
    }

    private func tickRound() {
        let remaining = uiState.secondsRemaining - 1
        if remaining == Self.tenSecondWarning {
            soundEngine.playTenSecondClapper()
            uiState.secondsRemaining = remaining
        } else if remaining <= 0 {
            soundEngine.playRoundEndHorn()
            advancePastRoundEnd()
        } else {
            uiState.secondsRemaining = remaining
        }
    }

    private func tickRest() {
        let remaining = uiState.secondsRemaining - 1
        if remaining >= 1 && remaining <= Self.restCountdownBeeps {
            soundEngine.playCountdownBeep()
            uiState.secondsRemaining = remaining
        } else if remaining <= 0 {
            soundEngine.playFinalCountdownBeep()
            uiState.phase = .round
            uiState.currentRound += 1
            uiState.secondsRemaining = uiState.roundSeconds
        } else {
            uiState.secondsRemaining = remaining
        }
    }

    private func advancePastRoundEnd() {
        if uiState.currentRound < uiState.totalRounds {
            uiState.phase = .rest
            uiState.secondsRemaining = uiState.restSeconds
        } else if uiState.preset?.hasGoldenScore == true {
            tickTask?.cancel()
            uiState.phase = .awaitingGoldenScoreDecision
            uiState.isRunning = false
        } else {
            tickTask?.cancel()
            uiState.phase = .finished
            uiState.isRunning = false
        }
    }

    deinit {
        tickTask?.cancel()
    }
}

File: app/src/main/java/com/saee/combattimer/TimerViewModel.kt
package com.saee.combattimer

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.saee.combattimer.engine.SoundEngine
import com.saee.combattimer.model.SportPreset
import com.saee.combattimer.model.TimerPhase
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

class TimerViewModel(application: Application) : AndroidViewModel(application) {

    private val soundEngine = SoundEngine(application)
    private var tickJob: Job? = null

    private val _uiState = MutableStateFlow(TimerUiState())
    val uiState: StateFlow<TimerUiState> = _uiState.asStateFlow()

    /** Loads a sport, optionally overriding its configurable fields, ready to start. */
    fun configure(preset: SportPreset, rounds: Int? = null, roundSeconds: Int? = null, restSeconds: Int? = null) {
        tickJob?.cancel()
        _uiState.value = TimerUiState(
            preset = preset,
            phase = TimerPhase.IDLE,
            totalRounds = rounds ?: preset.defaultRounds,
            currentRound = 1,
            roundSeconds = roundSeconds ?: preset.roundSeconds,
            restSeconds = restSeconds ?: preset.restSeconds,
            secondsRemaining = roundSeconds ?: preset.roundSeconds,
            isRunning = false
        )
    }

    fun start() {
        val current = _uiState.value
        if (current.phase == TimerPhase.IDLE) {
            _uiState.update { it.copy(phase = TimerPhase.ROUND, secondsRemaining = it.roundSeconds) }
        }
        runTicker()
    }

    fun pause() {
        tickJob?.cancel()
        _uiState.update { it.copy(isRunning = false) }
    }

    fun resume() {
        if (_uiState.value.phase in RESUMABLE_PHASES) {
            runTicker()
        }
    }

    fun reset() {
        val preset = _uiState.value.preset ?: return
        val rounds = _uiState.value.totalRounds
        val roundSeconds = _uiState.value.roundSeconds
        val restSeconds = _uiState.value.restSeconds
        configure(preset, rounds, roundSeconds, restSeconds)
    }

    /** Called from the Judo "match tied" prompt after the final round buzzer. */
    fun startGoldenScore() {
        tickJob?.cancel()
        _uiState.update { it.copy(phase = TimerPhase.GOLDEN_SCORE, secondsRemaining = 0, isRunning = true) }
        runTicker()
    }

    /** Ends the match without golden score (decision reached in regulation, or golden score scored). */
    fun finishMatch() {
        tickJob?.cancel()
        _uiState.update { it.copy(phase = TimerPhase.FINISHED, isRunning = false) }
    }

    private fun runTicker() {
        tickJob?.cancel()
        _uiState.update { it.copy(isRunning = true) }
        tickJob = viewModelScope.launch {
            while (isActive) {
                delay(1000)
                tick()
            }
        }
    }

    private fun tick() {
        when (_uiState.value.phase) {
            TimerPhase.ROUND -> tickRound()
            TimerPhase.REST -> tickRest()
            TimerPhase.GOLDEN_SCORE -> _uiState.update { it.copy(secondsRemaining = it.secondsRemaining + 1) }
            else -> Unit
        }
    }

    private fun tickRound() {
        val remaining = _uiState.value.secondsRemaining - 1
        when {
            remaining == TEN_SECOND_WARNING -> {
                soundEngine.playTenSecondClapper()
                _uiState.update { it.copy(secondsRemaining = remaining) }
            }
            remaining <= 0 -> {
                soundEngine.playRoundEndHorn()
                advancePastRoundEnd()
            }
            else -> _uiState.update { it.copy(secondsRemaining = remaining) }
        }
    }

    private fun tickRest() {
        val remaining = _uiState.value.secondsRemaining - 1
        when {
            remaining in 1..REST_COUNTDOWN_BEEPS -> {
                soundEngine.playCountdownBeep()
                _uiState.update { it.copy(secondsRemaining = remaining) }
            }
            remaining <= 0 -> {
                soundEngine.playFinalCountdownBeep()
                _uiState.update {
                    it.copy(
                        phase = TimerPhase.ROUND,
                        currentRound = it.currentRound + 1,
                        secondsRemaining = it.roundSeconds
                    )
                }
            }
            else -> _uiState.update { it.copy(secondsRemaining = remaining) }
        }
    }

    private fun advancePastRoundEnd() {
        val state = _uiState.value
        when {
            state.currentRound < state.totalRounds -> {
                _uiState.update { it.copy(phase = TimerPhase.REST, secondsRemaining = it.restSeconds) }
            }
            state.preset?.hasGoldenScore == true -> {
                tickJob?.cancel()
                _uiState.update { it.copy(phase = TimerPhase.AWAITING_GOLDEN_SCORE_DECISION, isRunning = false) }
            }
            else -> {
                tickJob?.cancel()
                _uiState.update { it.copy(phase = TimerPhase.FINISHED, isRunning = false) }
            }
        }
    }

    override fun onCleared() {
        tickJob?.cancel()
        soundEngine.release()
    }

    private companion object {
        const val TEN_SECOND_WARNING = 10
        const val REST_COUNTDOWN_BEEPS = 5
        val RESUMABLE_PHASES = setOf(TimerPhase.ROUND, TimerPhase.REST, TimerPhase.GOLDEN_SCORE)
    }
}
